qiskit/opflow/primitive_ops/primitive_op.py

```diff
@@ -246,7 +246,9 @@ def assign_parameters(self, param_dict: dict) -> OperatorBase:
                 return ListOp([self.assign_parameters(param_dict) for param_dict in unrolled_dict])
             if self.coeff.parameters <= set(unrolled_dict.keys()):
                 binds = {param: unrolled_dict[param] for param in self.coeff.parameters}
-                param_value = float(self.coeff.bind(binds))
+                param_value = complex(self.coeff.bind(binds))
+                if abs(param_value.imag) == 0:
+                    param_value = param_value.real
```

Comment:
I have seen cases where due to some math errors we have very small values say for an imaginary part e.g.  e-16 small. I guess one could use isclose() here with some very small tolerance to accommodate such. Or maybe we just leave it to caller to assure they truncate such values as needed.

Reply:
I think it would be better to leave this to the user, if they want to use the parameters in places where they have to be real they should make sure their input is valid 🙂

Reply:
Agree.